2022/python/Day7tree.py

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Dec 30 12:39:03 2022
+
+@author: raegoering
+
+--- Day 7: No Space Left On Device ---
+--- but with a tree ---
+
+"""
+from queue import Queue
+
+class Node:
+    def __init__(self, name, parent, size=0):
+        self.name = name
+        self.parent = parent
+        self.children = []
+        self.size = size
+
+    def __str__(self, level=0):
+        ret = "\t"*level+repr(self.name)+"\n"
+        for child in self.children:
+            ret += child.__str__(level+1)
+        return ret
+
+    def __repr__(self):
+        return self.name
+
+def make_tree(file):
+    root = Node("/", None)
+    for line in file:
+        line = line.rstrip()
+        if line.startswith("$ cd"):
+            cmd = line[5:]
+            if cmd == "/":
+                current_dir = root
+            elif cmd != "..":
+                child = None
+                for c in current_dir.children:
+                    if c.name == cmd:
+                        child = c
+                if child == None:
+                    child = Node(cmd, current_dir)
+                current_dir = child
+            elif cmd == "..":
+                current_dir = current_dir.parent
+        elif not line.startswith("$"):
+            if line.startswith("dir"):
+                current_dir.children.append(Node(line[4:], current_dir))
+            else:
+                size = line.split(sep=" ")[0]
```

Comment:
A fun thing you can do in Python is to assign to two variables from functions that return multiple things OR should I say mulTUPLE things 😆.

```
In [1]: x = "hello world"

In [2]: y,z = x.split(" ")

In [3]: y
Out[3]: 'hello'

In [4]: z
Out[4]: 'world'
```
Not a biggie, but it would fix having to call split here twice :D